core/hashfuncs.h

```diff
@@ -69,19 +69,52 @@ static inline uint32_t hash_djb2_one_32(uint32_t p_in,uint32_t p_prev=5381) {
 	return ((p_prev<<5)+p_prev)+p_in;
 }
 
+static inline uint32_t hash_one_uint64(const uint64_t p_int) {
+	uint64_t v=p_int;
+	v = (~v) + (v << 18); // v = (v << 18) - v - 1;
+	v = v ^ (v >> 31);
+	v = v * 21; // v = (v + (v << 2)) + (v << 4);
+	v = v ^ (v >> 11);
+	v = v + (v << 6);
+	v = v ^ (v >> 22);
+	return (int) v;
+}
+
 static inline uint32_t hash_djb2_one_float(float p_in,uint32_t p_prev=5381) {
 	union {
 		float f;
 		uint32_t i;
 	} u;
 
-	// handle -0 case
-	if (p_in==0.0f) u.f=0.0f;
-	else u.f=p_in;
+	// Normalize +/- 0.0 and NaN values so they hash the same.
+	if (p_in==0.0f)
+		u.f=0.0;
+	else if (isnan(p_in))
+		u.f=NAN;
```

Comment:
The convention so far has been to put these stuff (like isnan and NAN) under math_defs.h and math_funcs.h rather than using them from math.h. There is actually a PR for that #7787, can you check it?